### PR TITLE
FEATURE: Edit preview mode support for Neos 9

### LINKED ISSUE #23: TASK: 9.0 migrate `node.context.currentRenderingMode.edit`
`node.context` is not available anymore.

https://github.com/neos/neos-development-collection/blob/f4ed8369957209ba689cb2e1d9319e39716d6ac3/Neos.Neos/Resources/Private/Fusion/Prototypes/ContentCollection.fusion#L30

### LINKED ISSUE #3984: 9.0: Rebuild editPreviewModes
WIP writeup ...

currently the editPreviewMode is available via:
`node.context.currentRenderingMode` -> [UserInterfaceMode](https://github.com/neos/neos-development-collection/blob/30c8df283b4cbed355eda11bae111a7fb2ddb422/Neos.Neos/Classes/Domain/Model/UserInterfaceMode.php#L19)

and it will be added to the GlobalCacheIdentifiers

https://github.com/neos/neos-development-collection/blob/30c8df283b4cbed355eda11bae111a7fb2ddb422/Neos.Neos/Resources/Private/Fusion/Override/GlobalCacheIdentifiers.fusion#L9

The neos ui allows to change the preview modes, but this change affects the whole session.

-------

This has now been temporarily removed with the new CR, as `context` is dumped, and the nodeInfo helper (a replacement for `node.inBackend` https://github.com/neos/neos-ui/blob/f501f800f28c3523a90d52f807148cddf54ebb79/Classes/Fusion/Helper/NodeInfoHelper.php#L530) doesnt know anything about the preview mode too.

We should consider passing the currentRenderingMode via request parameter.
Question is, how we handle caching and allow uncached segments to work.

(fx. opening the preview in the iframe should not have the node uris converted, but when using the preview button (effectively the same route) the uris should function normally and be converted)

Im hoping to also fix these bugs in the same run:
- https://github.com/neos/neos-development-collection/issues/2347
- https://github.com/neos/neos-development-collection/issues/965

### LINKED ISSUE #4086: 9.0: Edit Preview Mode Support
Since the nodes have no .context anymore this cannot be used to determine the edit preview mode in neos 9.

PRs: 
- ~Neos: https://github.com/neos/neos-development-collection/pull/4067~
- ~Neos-UI: https://github.com/neos/neos-ui/pull/3411~
- ~Neos-Ui: 2 https://github.com/neos/neos-ui/pull/3571~
- Neos: https://github.com/neos/neos-development-collection/pull/4505
- Rector: https://github.com/neos/rector/pull/22

See also https://github.com/neos/neos-development-collection/issues/4396 for an explanation about why we opted in for removing the workspace based `inBackend`

### PR DESCRIPTION
**Pull request by mficzel**
The main change is that `edit` and `preview` action of the node controller are now separated with distinct constraints. This now allows for an mode to be edit and preview mode at the same time which is allowed by the configuration anyways.

- NodeController: has separate edit and preview actions that take the previewMode as argument
- NodeUriBuilder and LinkingService: will use preview / edit action once the main request is from the same action
- Neos.BackendHelper: has additional methods isEditMode, isPreviewMode and renderingModeCacheIdentifier

!!! The Neos.UI needs a change to provide the `editPreviewMode` parameter and call the edit or the preview action !!!

!!! This pr changes some behaviors slightly:

1. The distinction between edit and preview mode is now made based on the called action of the nodeController
2. The eel method `Neos.Node.inBackend(node)` is removed and is replaced with `Neos.Backend.isEditMode(request) || Neos.Backend.isPreviewMode(request)` this is non breaky as this method was never in an official release

Resolves: [#4086](https://github.com/neos/neos-development-collection/issues/4086)
Resolves partly: https://github.com/neos/neos-development-collection/issues/4396

**Upgrade instructions**

If fusion code relies on the edit preview mode beeing available via `node.context.currentRenderingMode` it has to be adjusted. Usually this means adjusting from code like `${node.context.currentRenderingMode.edit}` to `${Neos.Backend.isEditMode(request)}`.

**Review instructions**

Until the ui is adjusted this will always the preview as this was the existing method of the node controller. By setting the default value of the `editPreviewMode` on the `previewAction` to `rawContent` you can verify the rendering via different fusion pathes.

For now you can use the console to call the edit preview modes in your browser. You will have to replace the `node` with a node-address for your setup. You can then verify that the edit / preview mode is persisted during navigation after the initial call.

1. Open document in default preview mode `document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/preview?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677" `
2. Open document default edit mode `document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/edit?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677" `
3. Open document in rawContent mode (edit)
`document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/edit?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677&editPreviewMode=rawContent"` 

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

**Todos**
- [ ] required UI change https://github.com/neos/neos-ui/pull/3411
- [ ] `Neos.Backend.isEditMode` automated migration https://github.com/neos/rector/issues/21
- [ ] `Neos.Backend.isEditMode` migration in the UI https://github.com/neos/neos-ui/pull/3571

### CODE REVIEW

**Comment by ahaeslich** [comment]
@mficzel This will probably create a merge conflict with neos/neos-development-collection#4074.

**Comment by skurfuerst** [comment]
@mficzel @mhsdesign you let me know when we can review / merge this? :) all the best :)

**Comment by mficzel** [comment]
@akurfuerst a little rebase aside this should be ready to go. Only problem is the ui change that is needed here https://github.com/neos/neos-ui/pull/3411

@ahaeslich do you remember why we did not add a description

**Comment by mficzel** [comment]
@ahaeslich I added a description from memory. Did i miss something important.

I also merged the 9.0 branch into the state here. Looking OK so far and it seems to work fine for now.

**Comment by mficzel** [comment]
Noticed the related PRs:

- https://github.com/neos/neos-ui/pull/3471
- https://github.com/neos/neos-development-collection/pull/4217

**Comment by bwaidelich** [comment]
Any updates on this one?

**Comment by mficzel** [comment]
@bwaidelich i consider this ready backend-wise. However the ui will not work until it is adjusted to this change as it will default to preview mode if unadjusted.

The really important changes are for now:
1. The edit and preview are now distinct actions of the NodeController now  -> this idea taken from the pr https://github.com/neos/neos-development-collection/pull/4217 from @daniellienert 
2. The current fusion path (if specified) is set on the view directly instead of switching in the rootCase.
3. The backend helper has new methods `isEditMode`, `isPreviewMode` and `renderingModeCacheIdentifier` that each take the `request` as argument.
4. The LinkingService and NodeUriBuilder will also check the current request to create links with in the current editPreviewMode.

If we agree on those changes in general we should adjust the UI as follows: https://github.com/neos/neos-ui/pull/3411

1. The ui has to links to current document via the edit or the preview action (since before there was only the preview action one currently ends up in preview mode)
2. The ui edit preview mode selector has to adjust to the possibility of modes beeing edit- and previewMode at the same time. (the Settings allowed that anyways)
3. Investigate why new content in `rawContent` mode leads to reloading the document in `inPlace` mode

For now you can use the console to call the edit preview modes in your browser. You will have to replace the node with a node-address for your setup. You can then verify that the edit / preview mode is persisted during navigation after the initial call.

1. Open document in default preview mode `document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/preview?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677" `
2. Open document default edit mode `document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/edit?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677"`
3. Open document in rawContent mode (edit)
`document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/edit?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677&editPreviewMode=rawContent"`

File: Neos.Neos/Classes/Service/LinkingService.php [hunk 1]
```diff
@@ -377,7 +391,7 @@ public function createNodeUri(
             ->setArgumentsToBeExcludedFromQueryString($argumentsToBeExcludedFromQueryString)
             ->setFormat($format ?: $request->getFormat())
             ->setCreateAbsoluteUri($absolute)
-            ->uriFor($action, ['node' => $node], 'Frontend\Node', 'Neos.Neos');
+            ->uriFor($action, ['node' => $node], 'Frontend\Node', 'Neos.Neos') . '&grrr';
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Service/LinkingService.php]: grrrr

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php [hunk 2]
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
...
+        );
+    }
+
+    public function renderingModeCacheIdentifier(ActionRequest $request): string
+    {
+        if (
+            $request->getControllerPackageKey() === 'Neos.Neos'
+            && $request->getControllerName() === "Frontend\Node"
+            && ($request->getControllerActionName() === 'edit' || $request->getControllerActionName() === 'preview')
+        ) {
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]:
```suggestion
        if ($this->isEditMode($request) || $this->isPreviewMode($request)) {
```

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php [hunk 3]
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
...
+            $editPreviewModeArgument = $request->hasArgument('editPreviewMode') ? $request->getArgument('editPreviewMode') : null;
+            if (is_string($editPreviewModeArgument)) {
+                return $request->getControllerActionName() . ':' . $editPreviewModeArgument;
+            }
+            return $request->getControllerActionName();
+        } else {
+            return "show";
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]:
```suggestion
            return null;
```

instead?

File: Neos.Neos/Configuration/Settings.yaml [hunk 4]
```diff
@@ -278,7 +278,7 @@ Neos:
           title: Live
         inPlace:
           isEditingMode: true
-          isPreviewMode: false
+          isPreviewMode: true
```
Comment (mhsdesign) [inline Neos.Neos/Configuration/Settings.yaml:281]: Your thesis:

defaultEditPreviewMode should have both to true

Comment (mficzel) [reply Neos.Neos/Configuration/Settings.yaml:281]: One could rephrase this as the ui has to switch to edit / preview depending on what the defaultMode is.

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php (hunk 2 above)
Comment (mficzel) [reply Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]: This will check the controller and package twice ... not against it but also not a fan.

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php (hunk 3 above)
Comment (mficzel) [reply Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]: :-/ that would mean to change the return type to ?string ... i personally like cache identifier as string.

File: Neos.Neos/Configuration/Settings.yaml (hunk 4 above)
Comment (mficzel) [reply Neos.Neos/Configuration/Settings.yaml:281]: I agree this is not required anymore once the ui knows about the distinction of edit vs preview action. Currently this ensures that at least the preview mode is shown instead of an error.

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php [hunk 5]
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
         return $this->userService->getInterfaceLanguage();
     }
 
+    public function isEditMode(ActionRequest $request): bool
+    {
+        return ($request->getControllerPackageKey() === 'Neos.Neos'
+            && $request->getControllerName() === "Frontend\Node"
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]:
```suggestion
            && $request->getControllerName() === 'Frontend\\Node'
```

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php [hunk 6]
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
...
+            && $request->getControllerActionName() === 'edit'
+        );
+    }
+
+    public function isPreviewMode(ActionRequest $request): bool
+    {
+        return ($request->getControllerPackageKey() === 'Neos.Neos'
+            && $request->getControllerName() === "Frontend\Node"
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]:
```suggestion
            && $request->getControllerName() === 'Frontend\\Node'
```

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php (hunk 2 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]: I'm also in favor of preventing domain logic duplication.
`self::isEditMode(...) || self::isPreviewMode(...)` is much easier to comprehend and those string comparisons come with no remarkable overhead

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php (hunk 3 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]: I agree, nullable (return) types add complexity.
I'm just a bit unsure about the many magic strings in the code, but than again I think the actual value is not that important as long as it's unique, right?
nitpick: double quotes should be replaced though

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php [hunk 7]
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
         return $this->userService->getInterfaceLanguage();
     }
 
+    public function isEditMode(ActionRequest $request): bool
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]: This method could be static as far as I can tell
```suggestion
    public static function isEditMode(ActionRequest $request): bool
```

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php [hunk 8]
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
...
+    {
+        return ($request->getControllerPackageKey() === 'Neos.Neos'
+            && $request->getControllerName() === "Frontend\Node"
+            && $request->getControllerActionName() === 'edit'
+        );
+    }
+
+    public function isPreviewMode(ActionRequest $request): bool
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]:
```suggestion
    public static function isPreviewMode(ActionRequest $request): bool
```

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php [hunk 9]
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
...
+    {
+        return ($request->getControllerPackageKey() === 'Neos.Neos'
+            && $request->getControllerName() === "Frontend\Node"
+            && $request->getControllerActionName() === 'preview'
+        );
+    }
+
+    public function renderingModeCacheIdentifier(ActionRequest $request): string
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]:
```suggestion
    public static function renderingModeCacheIdentifier(ActionRequest $request): string
```

File: Neos.Neos/Classes/Service/LinkingService.php [hunk 10]
```diff
@@ -367,7 +369,19 @@ public function createNodeUri(
...
+        if (
+            $request->getControllerPackageKey() === 'Neos.Neos'
+            && $request->getControllerName() === "Frontend\Node"
+            && in_array($request->getControllerActionName(), ['edit', 'preview'])
+        ) {
+            $action = $request->getControllerActionName();
+            if ($request->hasArgument('editPreviewMode')) {
+                $arguments['editPreviewMode'] = $request->getArgument('editPreviewMode');
+            }
+        } else {
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Service/LinkingService.php]: We duplicate the logic here. Maybe something like this works:
```suggestion
        if (BackendHelper::isEditMode($request) || BackendHelper::isPreviewMode($request)) {
            $action = $request->getControllerActionName();
            if ($request->hasArgument('editPreviewMode')) {
                $arguments['editPreviewMode'] = $request->getArgument('editPreviewMode');
            }
        } else {
```

**Comment by mficzel** [comment]
I applied the suggested changes.

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php (hunk 3 above)
Comment (mficzel) [reply Neos.Neos/Classes/Fusion/Helper/BackendHelper.php]: Added single quotes as suggested.

**Comment by mhsdesign** [comment]
we should deprecate `Node.inBackend` as the concept from using the request is the correct one.

`node.workspace.isLive !== (!Node.inBackend)`

**Comment by grebaldi** [comment]
This PR should cover the content collection, right?:
https://github.com/neos/neos-development-collection/blob/f4ed8369957209ba689cb2e1d9319e39716d6ac3/Neos.Neos/Resources/Private/Fusion/Prototypes/ContentCollection.fusion#L30

(see: neos/rector#23)

**Comment by mficzel** [comment]
@grebaldi @mhsdesign good catch. Removed `Neos.Node.inBackend(node)` and replaced the usage with the `Neos.Backend.isEditMode(request) || Neos.Backend.isPreviewMode(request)`

**Comment by grebaldi** [comment]
Nice one @mficzel, thx!

I'll patch this into the test distribution over at https://github.com/neos/neos-ui/pull/3569 and get back to you with a full review :)

**Comment by mficzel** [comment]
@grebaldi zu needs this to get along with removed `Neos.Node.inBackend(node)` https://github.com/neos/neos-ui/pull/3571

**Comment by mficzel** [comment]
@grebaldi any support with adjusting the ui-side to the changes in the backend node entry points would be very much apreciated. Pretty sure the fact that this can be hardly tested is pert of the reason why there are no reviews yet.

**Comment by mhsdesign** [comment]
should `Neos.Backend.isPreviewMode` be true when you open the page externally?

image

(havent checked the behavior of this pr - just a reminder for me)

**Comment by mficzel** [comment]
> should Neos.Backend.isPreviewMode be true when you open the page externally?

@mhsdesign Pretty sure yes ... that was basically what the pr https://github.com/neos/neos-development-collection/pull/4217 was about that first tried to introduce separate edit and preview actions. Since previously the user session decided wether an edit or a preview mode was active it was hard to be in edit mode but render a preview using the show preview link.

**Comment by mhsdesign** [comment]
Im just wondering about caching.

When retrieving the pseudo `inBackend` information from the node (which actually only says if the node is not in the live workspace) caching works out of the box right?, because its a different node when `inBackend` changes.

But when taking the `request` we need to adjust the caching configuration so that this works?

```
renderer = ${Neos.Backend.isEditMode(request) ? 'edit' : (Neos.Backend.isPreviewMode(request) ? 'preview' : 'live')}
```

edit:

it seems you have done that in the global cache ids already:

```
renderingMode = ${Neos.Backend.renderingModeCacheIdentifier(request)}
```

File: Neos.Neos/Classes/Domain/Model/EditPreviewMode.php [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+final class EditPreviewMode
+{
+    protected function __construct(
+        public readonly string $name,
+        public readonly string $title,
+        public readonly ?string $fusionPath,
+        public readonly bool $isEditMode,
+        public readonly bool $isPreviewMode
+    ) {
+    }
```
Comment (grebaldi) [inline Neos.Neos/Classes/Domain/Model/EditPreviewMode.php:26]: We were talking about this model. It should contain a check, so that `$isEditMode` and `$isPreviewMode` cannot both be true at once (and that one of them must be true).

File: Neos.Neos/Classes/Domain/Repository/EditPreviewModeRepository.php [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        return EditPreviewMode::fromNameAndConfiguration($this->defaultEditPreviewMode, $this->editPreviewModeConfigurations[$this->defaultEditPreviewMode]);
+    }
+
+    public function findByName(string $name): EditPreviewMode
+    {
+        if (array_key_exists($name, $this->editPreviewModeConfigurations)) {
+            return EditPreviewMode::fromNameAndConfiguration($name, $this->editPreviewModeConfigurations[$name]);
+        }
+        throw new InvalidEditPreviewModeException(sprintf('"%s" is not a valid editPreviewMode', $name), 1683790077);
```
Comment (grebaldi) [inline Neos.Neos/Classes/Domain/Repository/EditPreviewModeRepository.php:42]:
```suggestion
        throw new InvalidEditPreviewModeException(sprintf('An editPreviewMode with the name "%s" does not exist', $name), 1683790077);
```

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php
Comment (grebaldi) [inline Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:1]: I'm very unsure whether it's wise to check for the current edit/preview mode via this EEL-Helper and by analyzing the request. As a consequence of this strategy, this is what needs to be done for out-of-band rendering in the UI:

https://github.com/neos/neos-ui/blob/0a858bb1cd5a31e4143f9b06144b0704c3acaad7/Classes/Domain/Model/Feedback/Operations/ReloadContentOutOfBand.php#L136-L146
*(These lines show the creation of a fake controller context, so this EEL-Helper is convinced to be executed in a different request than it actually is)*

Intuitively, I'd say it'd be better to just throw the current `EditPreviewMode` object into the global fusion context, so it can just be used like

```fusion
someKey.@if.inEditMode = ${editPreviewMode.isEditMode}
```

On the other hand, that fake controller context may not be the worst idea after all, because usually integrators wouldn't expect (and shouldn't notice) their code to go through such different execution paths.

However, semantically it seems very arbitrary to me to bind the current edit preview mode to a specific controller action...

File: Neos.Neos/Classes/Service/LinkingService.php [hunk 13]
```diff
@@ -367,7 +370,15 @@ public function createNodeUri(
         $request = $controllerContext->getRequest()->getMainRequest();
         $uriBuilder = clone $controllerContext->getUriBuilder();
         $uriBuilder->setRequest($request);
-        $action = $workspace && $workspace->isPublicWorkspace() && !$hiddenState->isHidden ? 'show' : 'preview';
+
+        if (BackendHelper::isEditMode($request) || BackendHelper::isPreviewMode($request)) {
```
Comment (grebaldi) [inline Neos.Neos/Classes/Service/LinkingService.php:374]: I'd prefer to move this closer to the domain. Looks like a job for the `EditPreviewModeRepository`:

```php
final class EditPreviewModeRepository
{
    public function findOneByActionRequest(ActionRequest $request): ?EditPreviewMode;
}
```

So that this check translates to 

```suggestion
        $editPreviewMode = $this->editPreviewModeRepository->findOneByActionRequest($request);
        if ($editPreviewMode) {
```

Depending on how we move forward with the `BackendHelper` itself, it should of course use the same method to retrieve the proper `EditPreviewMode` from the request.

File: Neos.Neos/Classes/Controller/Frontend/NodeController.php [hunk 14]
```diff
@@ -119,7 +129,38 @@ class NodeController extends ActionController
      * with unsafe requests from widgets or plugins that are rendered on the node
      * - For those the CSRF token is validated on the sub-request, so it is safe to be skipped here
      */
-    public function previewAction(string $node): void
+    public function previewAction(string $node, string $editPreviewMode = null): void
+    {
+        $editPreviewModeObject = $editPreviewMode ? $this->editPreviewModeRepository->findByName($editPreviewMode) : $this->editPreviewModeRepository->findDefault();
+        if ($editPreviewModeObject->isPreviewMode === false) {
+            throw new InvalidEditPreviewModeException(sprintf('"%s" is not a preview mode', $editPreviewMode), 1683127314);
```
Comment (grebaldi) [inline Neos.Neos/Classes/Controller/Frontend/NodeController.php:136]: I would also prefer to move this logic closer to the domain, by letting the `EditPreviewModeRepository` provide more specific retrieval methods:

```php
final class EditPreviewModeRepository
{
    /** @throws NotAnEditModeException */
    public function findOneEditModeByName(string $name): ?EditPreviewMode;

    /** @throws DefaultEditModeNotFoundException */
    public function findDefaultEditMode(): EditPreviewMode;

    /** @throws NotAPreviewModeException */
    public function findOnePreviewModeByName(string $name): ?EditPreviewMode;

    /** @throws DefaultPreviewModeNotFoundException */
    public function findDefaultPreviewMode(): EditPreviewMode;
}
```

File: Neos.Neos/Classes/Controller/Frontend/NodeController.php [hunk 15]
```diff
@@ -119,7 +129,38 @@ class NodeController extends ActionController
...
+     * @throws \Neos\Neos\Exception
+     * @Flow\SkipCsrfProtection We need to skip CSRF protection here because this action could be called
+     * with unsafe requests from widgets or plugins that are rendered on the node
+     * - For those the CSRF token is validated on the sub-request, so it is safe to be skipped here
+     */
+    public function editAction(string $node, ?string $editPreviewMode = null): void
+    {
+        $editPreviewModeObject = $editPreviewMode ? $this->editPreviewModeRepository->findByName($editPreviewMode) : $this->editPreviewModeRepository->findDefault();
+        if ($editPreviewModeObject->isEditMode === false) {
+            throw new InvalidEditPreviewModeException(sprintf('"%s" is not an edit mode', $editPreviewModeObject->name), 1683127295);
```
Comment (grebaldi) [inline Neos.Neos/Classes/Controller/Frontend/NodeController.php:158]: Likewise

**Review by grebaldi — COMMENTED** [review]
Hi @mficzel,

thanks for taking care of this :) I was able to successfully test this change in the scope of https://github.com/neos/neos-ui/pull/3569. 

Therein I stumbled upon some conceptual questions that we've already talked about briefly. The major questions are imho:

- Should it be possible for an `EditPreviewMode` to be both `edit`- AND `preview`-mode
  - Within this PR this is currently the default for `inPlace`
  - As we've spoken about this, my impression was that we generally agree that this shouldn't be possible, but that we need a solution for the UI to behave correctly, because of the change of endpoints
- Should the current `EditPreviewMode` be really dependent on the current action request?
  - The alternative would be to put the current `EditPreviewMode` object into the global fusion context
  - There are pros and cons for both solutions

All other comments I left are just some opinionated suggestions :)

So, in conclusion, we need to clear up those questions and then this should be good to go imho. But we need to coordinate closely, because once this is merged, the UI will cease to work, unless https://github.com/neos/neos-ui/pull/3569 is merged as well.

File: Neos.Neos/Classes/Domain/Model/EditPreviewMode.php (hunk 11 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/Domain/Model/EditPreviewMode.php:26]: dont say enum ;)

Comment (grebaldi) [reply Neos.Neos/Classes/Domain/Model/EditPreviewMode.php:26]: I didn't :D

But yeah, that would probably be better.

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php
Comment (mhsdesign) [reply Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:1]: Yes i rather have this additional context too. I remember that we discussed this at the Dresden sprint, but the discussion came up that in fusion we can't currently add new context variables because users would have to account for them in caching.

There is a way though: We need to pass the context variable behind fusions back like `request` so its always available in eel.

I will prepare a draft pr to show you what i mean ;)

**Comment by mficzel** [comment]
Am AFK for the next three weeks. Feel free to add all the changes that are necessary to this pr.

File: Neos.Neos/Classes/Fusion/Helper/BackendHelper.php
Comment (mhsdesign) [reply Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:1]: @grebaldi i prepared a concept over here: https://github.com/neos/neos-development-collection/pull/4425

this way we open the api of the runtime to allow other variables like `request` who are always present (in the instance of the runtime) and thus not affected by caching.

This would allow us to add a global `editPreviewMode` object with no dependency on the request. ;)

Comment (mhsdesign) [reply Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:1]:
```php
new Runtime(
  defaultContextVariables: [
      'request' => $request,
      'Neos' => [
        'UiMode' => new UiModeHelper($isBackend)
      ]
  ]
)
```

**Comment by mficzel** [comment]
Close in favor of https://github.com/neos/neos-development-collection/pull/4505 please let the branch live for a bit in case it is needed as comparison.
